### ISSUE TITLE
Question on candidate_sampling_probability

### ISSUE THREAD

**Issue by ssridhar2802** [tensorflow/recommenders#440]
First of all I wanted to thank @maciejkula and the entire TFRS team for this awesome project!  (not to mention the entire user community for posting a lot of great questions and answers I have learnt from)

I'm trying to use candidate_sampling_probability for bias correction and I'm passing `num_times_candidate_appears_in_interactions / num_all_interactions` to the retrieval task.  I have been following some of the related threads - https://github.com/tensorflow/recommenders/issues/257, https://github.com/tensorflow/recommenders/issues/232, https://github.com/tensorflow/recommenders/issues/140 and the paper - https://research.google/pubs/pub48840/   

My source of confusion is on how this correction will improve 

> popular items are overly penalized as negatives due to the high probability of being included in a batch

because in my mind it seems to be doing the opposite.  To piggy back on the example from @patrickorlando in https://github.com/tensorflow/recommenders/issues/257,

```
    [[ 7.9, 5.0, 6.7 ]

S = [ 6.8, 7.3, 5.7 ]

    [ 4.1, 3.8, 8.4 ]] 


P = [[0.005, 0.07, 0.1]]

Ln(P) = [[-5.3, -2.6, -2.3 ]]

S' = S - Ln(P)

   [[ 7.9 + 5.3, 5.0 + 2.6, 6.7 + 2.3 ]

  = [ 6.8 + 5.3, 7.3 + 2.6, 5.7 + 2.3 ]

   [ 4.1 + 5.3, 3.8 + 2.6, 8.4 + 2.3 ]]  
```

In this example: Probability of content id 1 (lower frequency - candidate sampling probability 0.005) seems to get boosted:

before correction:
`7.9 / (7.9+ 5.0+6.7) = 0.40306122449`

after correction:

`(7.9 + 5.3) / ((7.9 + 5.3) + (5.0 + 2.6) + (6.7 + 2.3)) = 0.44295302013`

Probability of content id 3 (higher frequency - candidate sampling probability 0.1) seems to be lowered after the correction.
before correction

`6.7 / (7.9+ 5.0+6.7) = 0.3418`

after correction

`(6.7+2.3) / ((7.9+ 5.3) +  (5.0+2.6) + (6.7+2.3)) = 0.3020
`

Maybe I'm misunderstanding something but doesn't this penalize popular content more?   I would really appreciate your insight and help!

Thanks,
Sridhar

**Comment by ssridhar2802** [comment]
I noticed that adding the candidate_sampling_probability also led to lower performance in the validation and test set.

**Comment by patrickorlando** [comment]
hey @ssridhar2802,

You can think of it this way.

Following the example above, let's assume the model has finished training.
The score for the pair (u1, i3) *with bias correction* is `0.3020`. During inference the correction is not applied, and the score will be `0.3418`. 

Popular items are given a boost at inference time by giving them a penalty during training.

**Comment by ssridhar2802** [comment]
@patrickorlando Thank you so much for clarifying,  It's clear to me now!
